src/librustc_mir/interpret/eval_context.rs

```diff
@@ -252,17 +252,22 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tc
         &self,
         substs: T,
     ) -> EvalResult<'tcx, T> {
+        if !substs.needs_subst() {
+            return Ok(self.tcx.erase_regions(&substs));
+        }
+        // Stack frames beyond the first stackframe have their own set of substs and it would be
+        // absolutely wrong to use them for any further substitutions. So we just yield the substs
+        // after region erasing and let the caller figure out if that is good enough for them.
+        if self.stack.len() > 1 {
+            return Ok(self.tcx.erase_regions(&substs));
+        }
         match self.stack.last() {
             Some(frame) => Ok(self.tcx.subst_and_normalize_erasing_regions(
                 frame.instance.substs,
                 self.param_env,
                 &substs,
             )),
-            None => if substs.needs_subst() {
-                err!(TooGeneric).into()
-            } else {
-                Ok(substs)
-            },
+            None => err!(TooGeneric).into(),
```

Comment:
New rounds of reviews: this should never happen (see other comments as to why).

But also (most/all?) `TooGeneric` errors should come from `LayoutError::Unknown` (which you should either turn into `TooGeneric` or handle *like* `TooGeneric` everywhere).

Reply:
There's also the problem that if we call `tcx.normalize_and_erase_regions`, at some point we'll end up hitting this assertion: https://github.com/rust-lang/rust/blob/91fa898975609f09f33e3e03fa8706c31a6b7d9b/src/librustc/traits/query/normalize_erasing_regions.rs#L59 So I've gated various calls that will transitively call `normalize_and_erase_regions` with an `if value.needs_subst() { return err!(TooGeneric); }`. I'm not sure how important that assertion is or why it's there. I checked the PR that created it, but it didn't explain the assert.